sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusHostBuilderExtensions.cs

```diff
@@ -93,5 +95,25 @@ public static IWebJobsBuilder AddServiceBus(this IWebJobsBuilder builder, Action
             builder.Services.AddSingleton<ServiceBusClientFactory>();
             return builder;
         }
+
+        public static IWebJobsBuilder AddServiceBusScaleForTrigger(this IWebJobsBuilder builder, TriggerMetadata triggerMetadata)
+        {
+            IServiceProvider serviceProvider = null;
+            Lazy<ServiceBusScalerProvider> scalerProvider = new Lazy<ServiceBusScalerProvider>(() => new ServiceBusScalerProvider(serviceProvider, triggerMetadata));
+
+            builder.Services.AddSingleton<IScaleMonitorProvider>(resolvedServiceProvider =>
+            {
+                serviceProvider = serviceProvider ?? resolvedServiceProvider;
```

Comment:
Can you clarify why we are doing this? Are we attempting to cache the first captured provider? Is this thread-safe? Why do we need to do this anyway - it doesn't seem to be used anywhere?

Reply:
We need only one `ServiceBusScalerProvider` to add `IScaleMonitorProvider` and `ITargetScalerProvider` to DI container. That's why I am using `Lazy<ServiceBusScalerProvider>` as it's [thread safe.](https://learn.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization#thread-safe-initialization)

We need to call [AddServiceBusScaleForTrigger ](https://github.com/Azure/azure-sdk-for-net/pull/34557/files#diff-2f37d5be90529dcbb846e2db09fd6d906bec85a72a6f637e8fd8d429d017c4aeR115 ) per trigger to use the scale entities associated with the trigger in `ScaleController`.

Reply:
Sorry my question was about the caching of the serviceProvider and using that in both of the AddSingleton lamdas.

Reply:
https://github.com/Azure/azure-sdk-for-net/pull/34557/files#diff-3b4ac6f1953372c2cb8f8861cda043dc9fe76aee45653789a18c54c1c3c2bf85R106
and
https://github.com/Azure/azure-sdk-for-net/pull/34557/files#diff-3b4ac6f1953372c2cb8f8861cda043dc9fe76aee45653789a18c54c1c3c2bf85R112

Reply:
We do not want to create all the instances twice so we are using one instance of `ServiceBusScalerProvider `
https://github.com/Azure/azure-sdk-for-net/pull/34557/files#diff-6e49cf486d47dbf5f4e52aa9911bbe8dba4a2c286bcfdfdebe10b84346fe3d11R27